File: sandbox-runner/patterns/delayed-callback/java/src/main/java/Activities.java
import io.temporal.activity.ActivityInterface;

@ActivityInterface
public interface Activities {
    String processPayment(Shared.PaymentPayload payload);

    String sendWebhookCallback(Shared.CallbackInput input);

    final class Impl implements Activities {
        @Override
        public String processPayment(Shared.PaymentPayload payload) {
            System.out.printf("Processing payment %s for $%.2f%n", payload.paymentId(), payload.amount());
            return String.format("Payment %s processed successfully for $%.2f",
                    payload.paymentId(), payload.amount());
        }

        @Override
        public String sendWebhookCallback(Shared.CallbackInput input) {
            // In production this would POST to input.callbackUrl() with input.payload() as the body.
            // For this demo we skip the actual HTTP call.
            System.out.printf("[stub] Would POST to %s with payload: %s%n",
                    input.callbackUrl(), input.payload());
            return String.format("Callback to %s delivered (stubbed)", input.callbackUrl());
        }
    }
}

File: sandbox-runner/patterns/delayed-callback/python/activities.py
import json
import urllib.request

from temporalio import activity

from shared import CallbackInput, PaymentPayload


@activity.defn
async def process_payment(payload: PaymentPayload) -> str:
    activity.logger.info(f"Processing payment {payload.payment_id} for ${payload.amount:.2f}")
    return f"Payment {payload.payment_id} processed successfully for ${payload.amount:.2f}"


@activity.defn
async def send_webhook_callback(input: CallbackInput) -> str:
    # In production this would POST to input.callback_url with input.payload as the body.
    # For this demo we skip the actual HTTP call.
    activity.logger.info(
        f"[stub] Would POST to {input.callback_url} with payload: {input.payload}"
    )
    return f"Callback to {input.callback_url} delivered (stubbed)"
